e2e/tests/core/client_test.go

```diff
@@ -119,3 +138,207 @@ func (s *ClientTestSuite) TestClientUpdateProposal_Succeeds() {
+
+		t.Run("get validators at latest height", func(t *testing.T) {
+			validators, err = s.GetValidatorSetByHeight(ctx, chainB, latestHeight.GetRevisionHeight())
+			s.Require().NoError(err)
+		})
+
+		t.Run("extract validator private keys", func(t *testing.T) {
+			privateKeys := s.extractChainPrivateKeys(ctx, chainB)
+			for i, pv := range privateKeys {
+				pubKey, err := pv.GetPubKey()
+				s.Require().NoError(err)
+
+				validator := tmtypes.NewValidator(pubKey, validators[i].VotingPower)
+
+				validatorSet = append(validatorSet, validator)
+				signers = append(signers, pv)
+			}
+		})
+	})
+
+	t.Run("create malicious header", func(t *testing.T) {
+		valSet := tmtypes.NewValidatorSet(validatorSet)
+		maliciousHeader, err = createMaliciousTMHeader(chainB.Config().ChainID, int64(latestHeight.GetRevisionHeight()), trustedHeight,
+			block.Header.GetTime().Add(time.Minute), valSet, valSet, signers, &block.Header)
+		s.Require().NoError(err)
+	})
+
+	t.Run("update client with duplicate misbehaviour header", func(t *testing.T) {
+		rlyWallet := s.CreateUserOnChainA(ctx, testvalues.StartingTokenAmount)
+		msgUpdateClient, err := clienttypes.NewMsgUpdateClient(ibctesting.FirstClientID, maliciousHeader, rlyWallet.FormattedAddress())
+		s.Require().NoError(err)
+
+		txResp, err := s.BroadcastMessages(ctx, chainA, rlyWallet, msgUpdateClient)
+		s.Require().NoError(err)
+		s.AssertValidTxResponse(txResp)
+	})
+
+	t.Run("ensure client status is frozen", func(t *testing.T) {
+		status, err := s.QueryClientStatus(ctx, chainA, ibctesting.FirstClientID)
+		s.Require().NoError(err)
+		s.Require().Equal(ibcexported.Frozen.String(), status)
+	})
+}
+
+// extractChainPrivateKeys returns a slice of tmtypes.PrivValidator which hold the private keys for all validator
+// nodes for a given chain.
+func (s *ClientTestSuite) extractChainPrivateKeys(ctx context.Context, chain *cosmos.CosmosChain) []tmtypes.PrivValidator {
+	testContainers, err := dockerutil.GetTestContainers(s.T(), ctx, s.DockerClient)
+	s.Require().NoError(err)
+
+	var filePvs []privval.FilePVKey
+	var pvs []tmtypes.PrivValidator
+	for _, container := range testContainers {
+		isNodeForDifferentChain := !strings.Contains(container.Names[0], chain.Config().ChainID)
+		isFullNode := strings.Contains(container.Names[0], fmt.Sprintf("%s-fn", chain.Config().ChainID))
+		if isNodeForDifferentChain || isFullNode {
+			continue
+		}
+
+		validatorPrivKey := fmt.Sprintf("/var/cosmos-chain/%s/config/priv_validator_key.json", chain.Config().Name)
+		privKeyFileContents, err := dockerutil.GetFileContentsFromContainer(ctx, s.DockerClient, container.ID, validatorPrivKey)
+		s.Require().NoError(err)
+
+		var filePV privval.FilePVKey
+		err = tmjson.Unmarshal(privKeyFileContents, &filePV)
+		s.Require().NoError(err)
+		filePvs = append(filePvs, filePV)
+	}
+
+	// we sort by address so that when we iterate through the ibcmock.PV later on they are in sync with the validators
+	// that are also sorted by address.
```

Comment:
Might be more clear to state that the private key slice needs to have the same ordering as the ibcmock.PV slice?

Reply:
actually this is not quite the case, the main thing is that the ibcmock.PV slice needs to be in the same order as the [validators](https://github.com/cosmos/ibc-go/blob/ad79f3dd88241890a06fb5e7ff627c4a3c266a53/e2e/tests/core/client_test.go#L203). I think the comment is still not quite clear though. Let me see if I can improve it.